### PR TITLE
IGNITE-8801 Forbid operations with atomic caches in transactions by default

### PR DESCRIPTION
**Pull request by J-Bakuli**
Need to change default behaviour of atomic operations to fail inside transactions.

1) Remove IGNITE_ALLOW_ATOMIC_OPS_IN_TX system property.
2) Set default value to restrict atomic operations in CacheOperationContext constructor without arguments and arguments for calls of another constructor.
3) Fix javadocs.

Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Comment by J-Bakuli** [comment]
@timoninmaxim Hi Maxim, could you please review my changes?

**Comment by J-Bakuli** [comment]
@timoninmaxim could you please review?

File: modules/core/src/main/java/org/apache/ignite/IgniteCache.java [hunk 1]
```diff
@@ -244,9 +244,9 @@
     public <K1, V1> IgniteCache<K1, V1> withKeepBinary();
 
     /**
-     * By default atomic operations are allowed in transaction.
-     * To restrict transactions from operations with atomic caches you can set system property
-     * {@link IgniteSystemProperties#IGNITE_ALLOW_ATOMIC_OPS_IN_TX IGNITE_ALLOW_ATOMIC_OPS_IN_TX} to {@code false}.
+     * Since 2.15.0 by default atomic operations inside transactions are not allowed.
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/IgniteCache.java:247]: "... are not allowed **by default**".

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java [hunk 2]
```diff
@@ -68,15 +76,15 @@ public class CacheOperationContext implements Serializable {
     /**
      * Constructor with default values.
      */
-    public CacheOperationContext() {
+    public CacheOperationContext() throws CacheException {
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java:79]: No need `throws` here.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java [hunk 3]
```diff
@@ -31,10 +32,17 @@
  * Cache operation context.
  */
 public class CacheOperationContext implements Serializable {
-    /** */
-    //TODO IGNITE-8801 remove this and set default as `false`.
-    public static final boolean DFLT_ALLOW_ATOMIC_OPS_IN_TX =
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java:36]: Let's keep this variable:
``
private static final boolean DFLT_ALLOW_ATOMIC_OPS_IN_TX = false;

public static final boolean defaultAllowAtomicOpsInTx() {
     return IgniteSystemProperties.getBoolean(IGNITE_ALLOW_ATOMIC_OPS_IN_TX, DFLT_ALLOW_ATOMIC_OPS_IN_TX);
}

```

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java [hunk 4]
```diff
@@ -31,10 +32,17 @@
...
-        IgniteSystemProperties.getBoolean(IGNITE_ALLOW_ATOMIC_OPS_IN_TX, true);
+    /**
+     * Since 2.15.0 atomic operations inside transactions are not allowed.
+     * To return the previous behaviour and to allow transactions in operations with atomic caches you can set system property
+     * {@link IgniteSystemProperties#IGNITE_ALLOW_ATOMIC_OPS_IN_TX IGNITE_ALLOW_ATOMIC_OPS_IN_TX} to {@code true}.
+     * <p>
+     * If you want to use atomic operations inside transactions in case they are restricted by system property,
+     * you should allow it before transaction start.
+     */
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java:42]: no need such a doc here, it's enough to have it only in `IgniteCache`.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java [hunk 5]
```diff
@@ -58,44 +61,76 @@ public class AtomicOperationsInTxTest extends GridCommonAbstractTest {
...
+     * {@link IgniteSystemProperties#IGNITE_ALLOW_ATOMIC_OPS_IN_TX IGNITE_ALLOW_ATOMIC_OPS_IN_TX} is changed to
+     * {@code true} before transaction start. Since 2.15.0 the default behaviour is set to {@code false}.
      * @throws Exception If failed.
      */
     @Test
-    public void testNotAllowedAtomicOperations() throws Exception {
-        checkOperations(false);
+    @WithSystemProperty(key = IGNITE_ALLOW_ATOMIC_OPS_IN_TX, value = "true")
+    public void testSetOfAtomicOperationsWithinTransactions() throws Exception {
+        checkOperations(true);
```
Comment (timoninmaxim) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java:129]: Let's also check operations with flag `false`.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java (hunk 2 above)
Comment (J-Bakuli) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java:79]: Thank you, I will remove

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java (hunk 4 above)
Comment (J-Bakuli) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java:42]: thank you, I will delete

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java (hunk 3 above)
Comment (J-Bakuli) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java:36]: thank you

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java (hunk 5 above)
Comment (J-Bakuli) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java:129]: 1 test with flag false is added (key=IGNITE_ALLOW_ATOMIC_OPS_IN_TX, value = "false") so now we have 2 tests with checkOperation() and flag "false" in SystemProperty

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java [hunk 6]
```diff
@@ -58,122 +61,186 @@ public class AtomicOperationsInTxTest extends GridCommonAbstractTest {
...
     }
 
     /**
+     * Tests whether allowing atomic cache operations before transaction starts does not cause exceptions.
+     * Since 2.15.0 the default behaviour is set to {@code false}.
+     * @throws Exception If failed.
+     */
+    @Test
+    @WithSystemProperty(key = IGNITE_ALLOW_ATOMIC_OPS_IN_TX, value = "true")
+    public void testAllowAtomicOperationsBeforeTransactionStarts() throws Exception {
```
Comment (timoninmaxim) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java:104]: This method doesn't throw Exceptions

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java [hunk 7]
```diff
@@ -58,122 +61,186 @@ public class AtomicOperationsInTxTest extends GridCommonAbstractTest {
     /** {@inheritDoc} */
     @Override protected void beforeTestsStarted() throws Exception {
         super.beforeTestsStarted();
+    }
 
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
```
Comment (timoninmaxim) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java:67]: Let's revert changes in beforeTest and afterTest, and start grid in `beforeTestsStarted` as it was before.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java [hunk 8]
```diff
@@ -58,122 +61,186 @@ public class AtomicOperationsInTxTest extends GridCommonAbstractTest {
...
 
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+        stopGrid(0);
+    }
+
     /**
+     * Tests whether enabling atomic cache operations within transactions is allowed.
+     * Since 2.15.0 the default behaviour is set to {@code false}.
```
Comment (timoninmaxim) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java:80]: No need to mention about 2.15 in those tests.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java [hunk 9]
```diff
@@ -32,9 +32,12 @@
  */
 public class CacheOperationContext implements Serializable {
     /** */
-    //TODO IGNITE-8801 remove this and set default as `false`.
-    public static final boolean DFLT_ALLOW_ATOMIC_OPS_IN_TX =
-        IgniteSystemProperties.getBoolean(IGNITE_ALLOW_ATOMIC_OPS_IN_TX, true);
+    public static final boolean DFLT_ALLOW_ATOMIC_OPS_IN_TX = false;
+
+    /** */
+    public static final boolean defaultAllowAtomicOpsInTx() {
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java:38]: Let's rename this method, as it doesn't return default value now

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java [hunk 10]
```diff
@@ -58,122 +61,186 @@ public class AtomicOperationsInTxTest extends GridCommonAbstractTest {
...
+        checkOperation(withAllowAtomicOpsInTx, cache -> cache.invokeAsync(1, new SetEntryProcessor()).get());
 
-        checkOperation(isAtomicCacheAllowedInTx, cache -> cache.invokeAll(map.keySet(), new SetEntryProcessor()));
+        checkOperation(withAllowAtomicOpsInTx, cache -> cache.invokeAll(map.keySet(), new SetEntryProcessor()));
 
-        checkOperation(isAtomicCacheAllowedInTx, cache -> cache.invokeAllAsync(map.keySet(),
+        checkOperation(withAllowAtomicOpsInTx, cache -> cache.invokeAllAsync(map.keySet(),
             new SetEntryProcessor()).get());
-
-        checkLock(isAtomicCacheAllowedInTx);
```
Comment (timoninmaxim) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java:165]: Why did you remove this line?

Comment (J-Bakuli) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java:165]: I moved it to checkOperation() and added err as a parameter:
checkLock(withAllowAtomicOpsInTx, err) to enable passing the test below:
`@Test
@WithSystemProperty(key = IGNITE_ALLOW_ATOMIC_OPS_IN_TX, value = "true")
public void testSetOfAtomicOperationsWithinTransactionsCheckFalseSystemPropertyTrue() throws Exception {
        checkOperations(false);
   }`

As a result, it became possible to write the following:
`else if (!withAllowAtomicOpsInTx && err == null) {
            cache = grid(0).cache(DEFAULT_CACHE_NAME);
            eCls = CacheException.class;
            eMsg = "Explicit lock can't be acquired within a transaction.";
        }`

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java (hunk 6 above)
Comment (J-Bakuli) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java:104]: thank you, I removed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java (hunk 9 above)
Comment (J-Bakuli) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheOperationContext.java:38]: thank you, I will rename it to allowAtomicOpsInTx()

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java (hunk 7 above)
Comment (J-Bakuli) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java:67]: Thank you, I will do it

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java (hunk 8 above)
Comment (J-Bakuli) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java:80]: Deleted mentioning, thank you

Comment (J-Bakuli) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java:80]: Deleted mentioning, thank you

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java (hunk 10 above)
Comment (J-Bakuli) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java:165]: Rolled back changes

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java (hunk 7 above)
Comment (J-Bakuli) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/AtomicOperationsInTxTest.java:67]: As was agreed later, beforeTest and afterTest are left.
beforeTestsStarted is removed

**Review by timoninmaxim — APPROVED** [review]
